Integrate permissions to Jenkins plugin

nodify-at opened the pull request:
## Hey, I just made a Pull Request!

I've integrated the permissions framework into Jenkins, so the users can not run "re-build" action for any kind of entities. A user is allowed to re-build only if he (or related group) owns the entity. 

I hope I created it in a correct way described as in RFC :)

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

mtlewis commented on plugins/jenkins-common/src/permissions.ts:
```diff
@@ -0,0 +1,34 @@
+  attributes: {
+    action: 'update',
+  },
+  resourceType: RESOURCE_TYPE_JENKINS,
```
Since this permission is authorized against catalog entities, we can use RESOURCE_TYPE_CATALOG_ENTITY here. Once we do that, the permission backend will automatically start using the catalog-backend to apply any conditions that need to be applied. This makes life a lot simpler - it means that we can remove all the changes to permission-backend in this PR, and everything should still work!

nodify-at replied:
I agree the only reason why I created the custom resource type and permissions is that in RFC has been described something like:

>  If your plugin needs authorization, you can define a set of permissions that describe the available actions. It’s recommended that you define all of the permissions for your plugin in a single module (see [Isomorphic packages for permissions](https://github.com/backstage/backstage/pull/9840#isomorphic-packages-for-permissions)).

I thought it could be better to have an independent permissions for a plugin so the client can able to choose or decide to activate or deactivate a permission inside for a specific plugin. (e.g. some of users would like to have re-build even if they enable catalog permissions to prevent deleting an entity) 

But I'm still trying to understand the permission plugin philosophy in backstage, if you think it should be fine to use catalog permissions here, I can remove/clean-up the code to use catalog resources.

mtlewis replied:
The permission definitely makes sense! It's the resource type, conditions, router etc. that seem to me to not be needed, as long as we want to authorize this permission in the context of catalog entities. 

If we wanted to instead consider jenkins jobs to be separate resources, and authorize permissions in the context of them instead of catalog entities, then we would need the stuff in permission-backend. The catalog entity approach seems like the right thing to me though :).

nodify-at replied:
Then I will use catalogEntityCreatePermission this permission request to check if user has a write access, is it fine for you? :)

mtlewis reviewed (commented):
Thanks for the contribution! Super impressive work given the permission framework is currently undocumented. Looking forward to getting this in!

nodify-at commented:
> Thanks for the contribution! Super impressive work given the permission framework is currently undocumented. Looking forward to getting this in!

Thanks for your feedback :)

nodify-at commented on plugins/jenkins/src/components/BuildsPage/lib/CITable/CITable.tsx:
```diff
@@ -174,6 +176,10 @@ const generatedColumns: TableColumn[] = [
     render: (row: Partial<Project>) => {
       const ActionWrapper = () => {
         const [isLoadingRebuild, setIsLoadingRebuild] = useState(false);
+        const { allowed, loading } = useEntityPermission(
```
You're right, I will provide necessary checks in the backend.

nodify-at commented:
@mtlewis  I improved the code as we discussed and added a check into backend (also added some tests) I'm not sure if we should keep conditionExports.ts since we use as a resource type catalog. But it might be good to provide it so the client can imageni what's needed for the permission check, do you a recommendation?

mtlewis commented on .changeset/ten-fireants-march.md:
```diff
@@ -0,0 +1,5 @@
+'@backstage/plugin-jenkins': minor
+---
+
+Integrated the permission plugin, if enabled, the actions can be executed only by owned user (or assigned to related group)
```
This isn't quite accurate - integrating with the permission policy puts control over who can perform actions in the hands of integrators. The specific users that will be permitted to perform the action will depend on what the integrator puts in their permission policy!

mtlewis commented on plugins/jenkins-backend/src/permissions/conditionExports.ts:
```diff
@@ -0,0 +1,27 @@
+import { createConditionExports } from '@backstage/plugin-permission-node';
+import { RESOURCE_TYPE_CATALOG_ENTITY } from '@backstage/plugin-catalog-common';
+
+const { conditions, createPolicyDecision } = createConditionExports({
```
We can actually rely on catalog-backend for these bits, and remove them entirely from here! Since the resourceType is `catalog-entity`, we'll be able to use all the rules from catalog-backend, and its permissionIntegrationRouter will be used to handle conditional responses. Hopefully makes life a lot easier!

mtlewis commented on plugins/jenkins-backend/src/permissions/rules/index.ts:
```diff
@@ -0,0 +1,30 @@
+ *
+ * @public
+ */
+export const jenkinsPermissionRules = {
```
Same goes for rules - using resourceType `catalog-entity` means jenkins-backend doesn't need its own rules at all - policy authors can use all the same ones from the catalog, so this module can be removed.

nodify-at replied:
You're right, we are using the catalog-entity and we don't need any additional rules inside permissions. I removed everything there.

mtlewis commented on plugins/jenkins-backend/src/service/router.ts:
```diff
@@ -14,24 +14,34 @@
 export interface RouterOptions {
   logger: Logger;
   jenkinsInfoProvider: JenkinsInfoProvider;
+  config?: Config;
+  discovery?: PluginEndpointDiscovery;
+  fetchApi?: { fetch: typeof fetch };
```
Perhaps I'm missing something, but I can't spot where these new parameters are being used!

nodify-at replied:
Sorry, it was my mistake, after removing integration router we don't need it any more :) I cleaned up the code and packages and removed also unused dependencies.

mtlewis reviewed (commented):
This is looking really great! Some nits and still a couple of parts that can be removed, but it's getting very close. Super excited about this PR :rocket:

nodify-at commented:
@mtlewis I've cleaned up the code and dependencies, improved the documentation to describe it with more detail :)

nodify-at commented:
While we're working on Jenkins plugin, Can we change the icon of rebuild and refresh? They use the same icon.

mtlewis commented on plugins/jenkins-backend/src/service/jenkinsApi.test.ts:
```diff
@@ -413,6 +423,34 @@ describe('JenkinsApi', () => {
+    ).rejects.toThrow(NotAllowedError);
+  });
+
+  it('buildProject should success if it have required permissions', async () => {
```
nit!
```suggestion
  it('buildProject should succeed if it have required permissions', async () => {
```

nodify-at replied:
ops :) thanks

mtlewis reviewed (approved):
Great stuff, thanks again for contributing this improvement!

nodify-at commented:
> While we're working on Jenkins plugin, Can we change the icon of rebuild and refresh? They use the same icon.



> While we're working on Jenkins plugin, Can we change the icon of rebuild and refresh? They use the same icon.

@mtlewis  can we use RotateLeftIcon for refresh? What do you think?

nodify-at commented:
> Great stuff, thanks again for contributing this improvement!

Thanks for your great support and feedback :) I understand the permissions now better.

mtlewis commented:
> > While we're working on Jenkins plugin, Can we change the icon of rebuild and refresh? They use the same icon.
> 
> @mtlewis can we use RotateLeftIcon for refresh? What do you think?

I'll leave this discussion for either the maintainers or  the codeowner of the jenkins plugin (@timja), since I'm not super familiar with it!

timja commented:
> I'll leave this discussion for either the maintainers or the codeowner of the jenkins plugin (@timja), since I'm not super familiar with it!

Feel free, but in a separate PR, and include before / after screen shots :)

nodify-at commented:
@timja  @mtlewis Can we take this change to the version which will be released today?

Rugvip commented on .changeset/gold-garlics-sing.md:
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-jenkins-common': major
```
Should be `minor`, so that we start out at `0.1.0`

Rugvip commented on plugins/jenkins-common/package.json:
```diff
@@ -0,0 +1,34 @@
+{
+  "name": "@backstage/plugin-jenkins-common",
+  "version": "0.1.0",
```
Should start at `0.0.0`

Rugvip commented on plugins/jenkins-backend/src/service/router.ts:
```diff
@@ -17,21 +17,25 @@
-import { Logger } from 'winston';
-import { JenkinsInfoProvider } from './jenkinsInfoProvider';
+import type { Logger } from 'winston';
+import type { JenkinsInfoProvider } from './jenkinsInfoProvider';
```
Should be no need to explicitly specify `type` imports, it's detected during transpilation based on usage within the file anyway

mtlewis commented:
I'll see if I can get some eyes on it!

Rugvip reviewed (commented):
Nice, thank you! 😁 

Just a few fixes needed for the setup of the new package, wouldn't want this one to be the first to reach 1.0 x)

Rugvip commented on .changeset/orange-cobras-shave.md:
```diff
@@ -0,0 +1,12 @@
+---
+'@backstage/plugin-jenkins-backend': minor
```
Not a breaking change in either of the existing packages right? Should be `patch` in that case

Rugvip commented:
Yep, can get it into today's release assuming some speedy turnaround of the feedback :grin:

nodify-at commented:
@Rugvip updated :) thanks

Rugvip reviewed (approved):
:shipit:

regicsolutions commented:
@nodify-at is there any documentation on how to enable this feature?

nodify-at commented:
@regicsolutions specific to this change, I didn't provide any documentation because this feature will be activated by default, if you enable the permissions api. You need to take a look on permission framework to enable it.

A short summary:

* Frontend: Permission check will be enable automatically and disables the button if you don't allow update access for the catalog resources. (It is recommended by backstage team to use this resource type, because Jenkins relates to an entity)

* Backend: To enable backend checks (which is recommended too) you need to provide permission api while creating the Jenkins router.

And you need to enable the permissions framework itself :)

Ping me in discord, if you need help.

@Rugvip  we need a documentation for this change? I can provide a summary if needed? If yes, can you please help me to find the correct place :)

Joonpark13 commented:
> And you need to enable the permissions framework itself :)

The docs for the permissions framework itself are currently in progress! They will hopefully be showing up on [backstage.io](https://backstage.io/docs) very soon.
